File: 249_Group_Shifted_Strings.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/group-shifted-strings/
# Given a string, we can "shift" each of its letter to its successive letter, for example: "abc" -> "bcd". We can
# keep "shifting" which forms the sequence: "abc" -> "bcd" -> ... -> "xyz"
# Given a list of strings which contains only lowercase alphabets, group all strings that belong to the same
# shifting sequence.

# For each string, shift so first character is "a". Map shifted string to list of all unshifted strings.
# Time - O(n), total number of chars of all strings
# Space - O(n)

from collections import defaultdict

class Solution(object):
    def groupStrings(self, strings):
        """
        :type strings: List[str]
        :rtype: List[List[str]]
        """
        shifted = defaultdict(list)

        for s in strings:

            shift = ord(s[0]) - ord('a')            # assumes s is not empty
            s_shifted = "".join([chr((ord(c) - ord('a') - shift) % 26 + ord('a')) for c in s])
            shifted[s_shifted].append(s)

        return shifted.values()

File: java/249_Group_Shifted_Strings.java
/*
https://leetcode.com/problems/group-shifted-strings/
Given a string, we can "shift" each of its letter to its successive letter, for example: "abc" -> "bcd".
We can keep "shifting" to form a sequence.
Given a list of strings which contains only lowercase alphabets, group all strings that belong to the same shifting sequence.

Shift each string so the rirst char is 'a'.  Map each such shifted string to a list of original unshifted strings.
Return the map values.
Time - O(n), where n is the total length in chars of all strings
Space - O(n)
*/

public class Solution {
    public List<List<String>> groupStrings(String[] strings) {

        List<List<String>> groups = new ArrayList<List<String>>();
        Map<String, List<String>> equivalents = new HashMap<String, List<String>>();

        for (String s : strings) {

            int shift = s.charAt(0) - 'a';                      // number of shifts to make first char 'a'
            StringBuilder shifted = new StringBuilder("a");     // first char is always 'a'

            for (int i = 1; i < s.length(); ++i) {              // shift each char
                int c = s.charAt(i) - shift;
                if (c < 'a')                                    // mod 26
                    c += 26;
                shifted.append((char) c);
            }

            String shiftString = shifted.toString();
            if (!equivalents.containsKey(shiftString))
                equivalents.put(shiftString, new ArrayList<String>());
            equivalents.get(shiftString).add(s);                // add to mapping

        }

        for (String shiftString : equivalents.keySet())
            groups.add(equivalents.get(shiftString));

        return groups;
    }
}
